Step1

imcool2551 opened the pull request.

imcool2551 committed:
feat: Answer

delete() 메서드에서
1) 권한 검사수행
2) 매개변수로 넘긴 DeleteHistories에 DeleteHistory 추가
를 수행했습니다.

imcool2551 committed:
feat: Answers

List<Answer> 를 래핑하는 일급 컬렉션 클래스입니다.

imcool2551 committed:
feat: DeleteHistories

List<DeleteHistory>를 래핑하는 일급 컬렉션입니다.

imcool2551 committed:
feat: Question

delete() 메서드에서
1) 권한 검사
2) DeleteHistory 추가
3) 1:N 관계를 맺는 Answer 타입의 delete 호출
을 수행했습니다.

imcool2551 committed:
refactor: QnAService

도메인 메서드로 로직 이동

imcool2551 committed:
test: AnswerTest

1) 다른 사람의 답변 삭제시 예외
2) 삭제 성공시 deleted 필드 변경과 DeleteHistory 생성
두 가지 경우를 테스트했습니다.

imcool2551 committed:
test: Question

1) 다른 사람 질문 삭제시 예외
2) 다른 사람 답변 존재시 삭제하면 예외
3) 삭제 성공시 deleted 필드 변경, question+answer 의 DeleteHistory 추가
세 가지 경우를 테스트했습니다.

yongdae commented on src/main/java/qna/service/QnAService.java:
```diff
@@ -26,33 +25,17 @@ public class QnAService {
+    public void deleteQuestion(User loginUser, long questionId) throws CannotDeleteException {
+        Question question = findQuestionById(questionId);
+        DeleteHistories deleteHistories = new DeleteHistories();
+        question.delete(loginUser, deleteHistories, now());
```
Question 객체의 delete() 메서드가 DeleteHistories를 반환하게 만들어보면 어떨까요?

yongdae commented on src/main/java/qna/domain/Question.java:
```diff
@@ -39,26 +38,21 @@ public Question(long id, String title, String contents) {
+    public void delete(User loginUser, DeleteHistories deleteHistories, LocalDateTime createDate) throws CannotDeleteException {
+        validateOwner(loginUser);
+        this.deleted = true;
+        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, getId(), getWriter(), createDate));
```
DeleteHistory 객체에 [정적 팩토리 메서드](https://7942yongdae.tistory.com/147)를 적용해서 사용해보면 좋을 것 같아요. :)

yongdae commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,34 @@
+    public Answers() {
+    }
+
+    public void delete(User loginUser, DeleteHistories deleteHistories, LocalDateTime createDate) throws CannotDeleteException {
```
여기서도 DeleteHistories를 매개변수로 받아서 쓰지 않고 DeleteHistories를 반환하게 만들어보면 좋을 것 같아요.

yongdae commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -43,29 +45,32 @@ public Answer(Long id, User writer, Question question, String contents) {
 
-    public boolean isDeleted() {
-        return deleted;
+    public void delete(User loginUser, DeleteHistories deleteHistories, LocalDateTime createDate) throws CannotDeleteException {
```
Question, Answer의 delete() 메서드는 매개변수로 LocalDateTime createDate를 사용하는데
외부에서 생성시간을 전달받아 DeleteHistory를 만들어야 할 이유가 있나요? 🤔

imcool2551 replied:
시간에 의존적인 요구사항이 추가 될 경우 (ex. 작성 후 1년이 지나면 삭제 금지)나 시간에 의존적인 테스트를 작성할 필요가 생긴다면 함수 내부에서 직접 now() 를 호출할 때 테스트가 어려워지기 때문에 미리 매개변수화했습니다. 혹시 그러한 필요가 생길 때 변경하고 현재 요구사항에는 내부에서 직접 생성해도 문제가 없으니 삭제하는것이 더 나은 선택인가요? 😭

yongdae replied:
미리 요구사항을 고민해서 반영하는게 좋지 않다고는 생각되지 않아요!
다만 미션에서는 리팩토링을 주제로 하고 있으니 현재 있는 코드를 개선하는 것에 중점을 두는게 어떨까요? 🤔

yongdae commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.List;
+
+@Embeddable
+public class Answers {
```
Answers에 대한 테스트도 추가해 볼 수 있지 않을까요?

yongdae reviewed (changes requested):
안녕하세요. 상욱님 ~ :)
마지막 볼링 미션을 함께 하게 된 리뷰어 권용대라고 합니다.

첫번째 단계인 질문 삭제하기 리팩토링 미션을 잘 진행해주셨네요. 👍 
소소하지만 진행해주신 내용에 대해 몇 가지 의견을 남겨 놓았습니다.

확인하시고 다시 리뷰 부탁드리겠습니다. :)

그럼 미션 마지막까지 잘 부탁드리고 의견이 있으시거나 고민이 있으시다면
언제든 코멘트나 DM을 통해 이야기 부탁드릴께요. 🙇

imcool2551 committed:
refactor: Answer

1. 매개변수로 받은 DeleteHistories에 DeleteHistory를 추가하는 대신 DeleteHistory를 반환하도록 했습니다.
2. DeleteHistory에 설정할 삭제 시간을 매개변수로 받는 대신 DeleteHistory가 생성될 때 현재 시간을 삭제 시간으로 설정하도록 변경했습니다.

테스트도 변경사항에 맞게 수정했습니다.

imcool2551 committed:
refactor+test: Answers

1. Answer와 비슷하게 DeleteHistories를 받는 대신 DeleteHistories를 반환하도록 했습니다.
2. Answer와 비슷하게 삭제 시간을 매개변수로 받는 대신 현재 시간을 삭제 시간으로 설정하도록 했습니다.

Answers에 포함된 Answer들이 DeleteHistories에 잘 담기는지 테스트했습니다.

imcool2551 committed:
refactor: QnaService

DeleteHistories를 생성해서 Question.delete() 에 매개변수로 넘기는 대신 반환값을 받도록 했습니다.

imcool2551 committed:
refactor: Question

Answer, Answers 리팩토링과 동일하게 DeleteHistories를 반환하도록 했습니다.
연관된 Answers를 DeleteHistories에 한 번에 편하게 담기위해 DeleteHistories에 addAll() 메서드를 추가했습니다.

imcool2551 committed:
refactor: DeleteHistory

생성자 대신 정적 팩터리 메서드를 사용했습니다. createdDate를 주입받는 대신 필드에서 현재 시간으로 초기화했습니다.

imcool2551 committed:
refactor: QnaServiceTest

yongdae commented on src/main/java/qna/service/QnAService.java:
```diff
@@ -26,33 +23,16 @@ public class QnAService {
+    @Transactional
+    public void deleteQuestion(User loginUser, long questionId) throws CannotDeleteException {
+        Question question = findQuestionById(questionId);
+        DeleteHistories deleteHistories = question.delete(loginUser);
```
의견 드린 부분을 잘 반영해주셨네요. 👍

yongdae commented on src/test/java/qna/domain/AnswersTest.java:
```diff
@@ -0,0 +1,27 @@
+import static qna.domain.QuestionTest.Q1;
+import static qna.domain.UserTest.JAVAJIGI;
+
+class AnswersTest {
```
Answers 테스트 뿐만 아니라
Question, Answer에 대한 테스트도 잘 만들어서 주셨네요. :)

yongdae reviewed (approved):
의견 드렸던 부분들을 빠르고 정확하게
그리고 커밋 메시지도 정말 멋지게 잘 작성해서 진행해주셨네요. 💯 

그럼 본격적인 볼링 미션을 위해 다음 단계로 넘어가도록 하겠습니다.
수고 많으셨습니다. 🙇
